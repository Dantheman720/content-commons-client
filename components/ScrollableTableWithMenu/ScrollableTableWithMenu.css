.ui.grid .row.items_tableMenus_wrapper {
  padding-bottom: 0; }
  @media screen and (max-width: 930px) {
    .ui.grid .row.items_tableMenus_wrapper {
      flex-direction: column-reverse; } }
  @media screen and (max-width: 767px) {
    .ui.grid .row.items_tableMenus_wrapper .column {
      padding-left: 0;
      padding-right: 0; } }
  .ui.grid .row.items_tableMenus_wrapper .column.items_tableMenus {
    display: flex; }
    @media screen and (min-width: 930px) {
      .ui.grid .row.items_tableMenus_wrapper .column.items_tableMenus {
        justify-content: flex-end;
        margin-bottom: 1em; } }
    @media screen and (max-width: 929px) {
      .ui.grid .row.items_tableMenus_wrapper .column.items_tableMenus {
        margin-bottom: 1em; } }

.items_table {
  position: relative;
  max-height: 500px;
  overflow: scroll;
  border: 1px solid rgba(34, 36, 38, 0.15);
  /* ONLY IN WORKS IN CHROME & SAFARI */ }
  @media screen and (max-width: 767px) {
    .items_table {
      border: none;
      max-height: 300px;
      overflow: auto; } }
  .items_table::-webkit-scrollbar {
    display: none; }
  .items_table_wrapper {
    padding-bottom: 0;
    margin-top: -1.5em !important; }
    @media screen and (max-width: 767px) {
      .items_table_wrapper {
        padding: 20px 0 0 0 !important;
        margin-top: -2em !important; } }
  .items_table table.ui.sortable.table {
    table-layout: fixed;
    width: 100%;
    margin-top: 0;
    border: none;
    border-radius: 0; }
    @media screen and (max-width: 767px) {
      .items_table table.ui.sortable.table thead {
        display: none; } }
    .items_table table.ui.sortable.table th {
      letter-spacing: 0.5px; }
    @media screen and (max-width: 767px) {
      .items_table table.ui.sortable.table tr {
        padding: 5px 3px;
        margin-bottom: 5px;
        border: 1px solid #f1f1f1;
        box-shadow: none !important; } }
    .items_table table.ui.sortable.table td:first-child, .items_table table.ui.sortable.table th:first-child {
      position: sticky;
      left: 0;
      width: 300px !important; }
    .items_table table.ui.sortable.table td {
      z-index: 1;
      background-color: #fff;
      box-shadow: 3px 0 3px 1px rgba(0, 0, 0, 0.1); }
      @media screen and (max-width: 767px) {
        .items_table table.ui.sortable.table td.items_table_item {
          position: relative;
          z-index: 2;
          width: auto !important;
          padding: 0 !important; }
          .items_table table.ui.sortable.table td.items_table_item:not(:first-child) {
            max-height: 0;
            height: auto;
            opacity: 0;
            transition: max-height 0.3s ease, opacity 0.2s ease; } }
      .items_table table.ui.sortable.table td .primary_col {
        display: flex; }
  .items_table_mobileHeader {
    display: inline-block;
    width: 60px;
    margin-right: 2em; }
    @media screen and (min-width: 768px) {
      .items_table_mobileHeader {
        display: none; } }
