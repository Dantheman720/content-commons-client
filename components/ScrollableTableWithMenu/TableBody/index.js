import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { Loader, Table } from 'semantic-ui-react';
import ApolloError from 'components/errors/ApolloError';
import MyProjectPrimaryCol from 'components/admin/Dashboard/MyProjects/MyProjectPrimaryCol';
import TableMobileDataToggleIcon from 'components/ScrollableTableWithMenu/TableMobileDataToggleIcon';
import { formatDate } from 'lib/utils';
import './TableBody.scss';

const TEAM_VIDEO_PROJECTS_QUERY = gql`
  query VideoProjectsByTeam(
    $team: String!, $first: Int, $skip: Int, $searchTerm: String
  ) {
    videoProjects(
      where: {
        AND: [
          { team: { name: $team } },
          {
            OR: [
              { projectTitle_contains: $searchTerm },
              { descPublic_contains: $searchTerm },
              { descInternal_contains: $searchTerm },
              { projectType_contains: $searchTerm },
              {
                categories_some: {
                  translations_some: { name_contains: $searchTerm }
                }
              },
              {
                author: {
                  OR: [
                    { firstName_contains: $searchTerm },
                    { lastName_contains: $searchTerm },
                    { email_contains: $searchTerm }
                  ]
                }
              }
            ]
          }
        ]
      },
      first: $first,
      skip: $skip
     ) {
      id
      createdAt
      updatedAt
      team {
        name
        organization
      }
      author {
        firstName
        lastName
      }
      projectTitle
      status
      visibility
      thumbnails {
        url
        alt
        caption
        filename
        filetype
        dimensions {
          width
          height
        }                
      }      
    }
  }
`;

const normalizeData = videoProjects => {
  const normalizedVideoProjects = [];

  videoProjects.forEach( videoProject => {
    const normalizedProject = Object.create( {}, {
      id: { value: videoProject.id },
      createdAt: { value: formatDate( videoProject.createdAt ) },
      updatedAt: { value: formatDate( videoProject.updatedAt ) },
      projectTitle: { value: videoProject.projectTitle },
      author: { value: `${videoProject.author ? videoProject.author.firstName : ''} ${videoProject.author ? videoProject.author.lastName : ''}` },
      team: { value: videoProject.team.name },
      status: { value: videoProject.status },
      visibility: { value: videoProject.visibility },
      thumbnail: {
        value: {
          url: videoProject.thumbnails[0].url,
          alt: videoProject.thumbnails[0].alt
        }
      }
    } );
    normalizedVideoProjects.push( normalizedProject );
  } );

  return normalizedVideoProjects;
};

const TableBody = props => {
  const {
    searchTerm,
    selectedItems,
    tableHeaders,
    toggleItemSelection,
    variables,
    windowWidth
  } = props;

  return (
    <Query
      query={ TEAM_VIDEO_PROJECTS_QUERY }
      variables={ { ...variables } }
    >
      { ( { loading, error, data: { videoProjects } } ) => {
        if ( loading ) {
          return (
            <Table.Body>
              <Table.Row>
                <Table.Cell>
                  <Loader active inline size="small" />
                  <span style={ { marginLeft: '0.5em', fontSize: '1.5em' } }>
                    Loading...
                  </span>
                </Table.Cell>
              </Table.Row>
            </Table.Body>
          );
        }
        if ( error ) {
          return (
            <Table.Body>
              <Table.Row>
                <Table.Cell>
                  <ApolloError error={ error } />
                </Table.Cell>
              </Table.Row>
            </Table.Body>
          );
        }
        if ( !videoProjects ) return null;

        if ( searchTerm && !videoProjects.length ) {
          return (
            <Table.Body>
              <Table.Row>
                <Table.Cell>
                  No results for &ldquo;{ searchTerm }&rdquo;
                </Table.Cell>
              </Table.Row>
            </Table.Body>
          );
        }

        const tableData = normalizeData( videoProjects );

        return (
          <Table.Body className="myProjects">
            { tableData.map( d => (
              <Table.Row
                key={ d.id }
                className={ d.isNew ? 'myProjects_newItem' : '' }
              >
                { tableHeaders.map( ( header, i ) => (
                  <Table.Cell
                    data-header={ header.label }
                    key={ `${d.id}_${header.name}` }
                    className="items_table_item"
                  >
                    { i === 0
                      ? (
                        // Table must include .primary_col div for fixed column
                        <Fragment>
                          <div className="primary_col">
                            <MyProjectPrimaryCol
                              d={ d }
                              header={ header }
                              selectedItems={ selectedItems }
                              toggleItemSelection={ toggleItemSelection }
                            />
                          </div>
                          { windowWidth && <TableMobileDataToggleIcon /> }
                        </Fragment>
                      )
                      : (
                        <Fragment>
                          <span>
                            <div className="items_table_mobileHeader">{ header.label }</div>
                            { d[header.name] }
                          </span>
                          <br />
                          { header.label === 'MODIFIED'
                            ? <span>{ d.status }</span>
                            : null }
                        </Fragment>
                      ) }
                  </Table.Cell>
                ) ) }
              </Table.Row>
            ) ) }
          </Table.Body>
        );
      } }
    </Query>
  );
};

TableBody.propTypes = {
  searchTerm: PropTypes.string,
  selectedItems: PropTypes.object,
  tableHeaders: PropTypes.array,
  toggleItemSelection: PropTypes.func,
  variables: PropTypes.object,
  windowWidth: PropTypes.oneOfType( [
    PropTypes.string,
    PropTypes.number
  ] )
};

export default TableBody;
export { TEAM_VIDEO_PROJECTS_QUERY };
