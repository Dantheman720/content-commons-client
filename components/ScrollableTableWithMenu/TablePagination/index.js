import React from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { Pagination } from 'semantic-ui-react';
import ApolloError from 'components/errors/ApolloError';
import './TablePagination.scss';

const TEAM_VIDEO_PROJECTS_COUNT_QUERY = gql`
  query VideoProjectsCountByTeam( $team: String!, $searchTerm: String ) {
    videoProjects(
      where: {
        AND: [
          { team: { name: $team } },
          {
            OR: [
              { projectTitle_contains: $searchTerm },
              { descPublic_contains: $searchTerm },
              { descInternal_contains: $searchTerm },
              {
                categories_some: {
                  translations_some: { name_contains: $searchTerm }
                }
              },
              {
                author: {
                  OR: [
                    { firstName_contains: $searchTerm },
                    { lastName_contains: $searchTerm },
                    { email_contains: $searchTerm }
                  ]
                }
              }
            ]
          }
        ]
      }
    ) {
      id
    }
  }
`;

const TablePagination = props => {
  const {
    activePage, handlePageChange, itemsPerPage, variables
  } = props;
  return (
    <Query
      query={ TEAM_VIDEO_PROJECTS_COUNT_QUERY }
      variables={ { ...variables } }
    >
      { ( { loading, error, data: { videoProjects } } ) => {
        if ( loading ) return 'Loading....';
        if ( error ) return <ApolloError error={ error } />;
        if ( !videoProjects ) return null;

        const projectsCount = videoProjects.length;
        const totalPages = Math.ceil( projectsCount / itemsPerPage );

        if ( projectsCount > 0 ) {
          return (
            <Pagination
              activePage={ activePage }
              totalPages={ totalPages }
              nextItem={ { content: 'Next >', disabled: activePage === totalPages } }
              prevItem={ { content: '< Previous', disabled: activePage === 1 } }
              siblingRange="2"
              firstItem={ null }
              lastItem={ null }
              onPageChange={ handlePageChange }
            />
          );
        }
        return null;
      } }
    </Query>
  );
};

TablePagination.propTypes = {
  activePage: PropTypes.number,
  handlePageChange: PropTypes.func,
  itemsPerPage: PropTypes.number,
  variables: PropTypes.object
};

export default TablePagination;
export { TEAM_VIDEO_PROJECTS_COUNT_QUERY };
