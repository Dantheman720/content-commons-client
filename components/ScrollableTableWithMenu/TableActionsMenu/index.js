/**
 *
 * TableActionsMenu
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { Button, Checkbox, Confirm } from 'semantic-ui-react';
import ApolloError from 'components/errors/ApolloError';
import editIcon from 'static/images/dashboard/edit.svg';
import createIcon from 'static/images/dashboard/create.svg';
import deleteIcon from 'static/images/dashboard/delete.svg';
import archiveIcon from 'static/images/dashboard/archive.svg';
import unpublishIcon from 'static/images/dashboard/unpublish.svg';
import ConfirmModalContent from 'components/admin/ConfirmModalContent/ConfirmModalContent';
import { TEAM_VIDEO_PROJECTS_QUERY } from '../TableBody';
import './TableActionsMenu.scss';

const DELETE_VIDEO_PROJECTS_MUTATION = gql`
  mutation DeleteManyVideoProjects($where: VideoProjectWhereInput) {
    deleteProjects: deleteManyVideoProjects(where: $where) {
      count
    }
  }
`;

const UNPUBLISH_VIDEO_PROJECTS_MUTATION = gql`
  mutation UnpublishManyVideoProjects(
    $data: VideoProjectUpdateManyMutationInput!,
    $where: VideoProjectWhereInput) {
    unpublish: updateManyVideoProjects(data: $data, where: $where) {
      count
    }
  }
`;

/* eslint-disable react/prefer-stateless-function */
class TableActionsMenu extends React.Component {
  state = {
    deleteConfirmOpen: false
  };

  handleUnpublish = unpublishFn => {
    const { selectedItems } = this.props;
    unpublishFn( {
      variables: {
        data: { status: 'DRAFT', visibility: 'INTERNAL' },
        where: {
          AND: [
            { id_in: [...selectedItems.keys()] },
            { status_not: 'DRAFT' }
          ]
        }
      }
    } );
  }

  handleDeleteCancel = () => {
    this.setState( { deleteConfirmOpen: false } );
  }

  handleDeleteConfirm = deleteFn => {
    const { queryVariables, selectedItems } = this.props;
    deleteFn( {
      variables: {
        where: { id_in: [...selectedItems.keys()] }
      },
      refetchQueries: [{
        query: TEAM_VIDEO_PROJECTS_QUERY,
        variables: { ...queryVariables }
      }]
    } );
  }

  displayConfirmDelete = () => {
    this.setState( { deleteConfirmOpen: true } );
  }

  render() {
    const { displayActionsMenu, toggleAllItemsSelection } = this.props;

    return (
      <div className="actionsMenu_wrapper">
        <Checkbox
          className={ displayActionsMenu ? 'actionsMenu_toggle actionsMenu_toggle--active' : 'actionsMenu_toggle' }
          onChange={ toggleAllItemsSelection }
        />
        <div className={ displayActionsMenu ? 'actionsMenu active' : 'actionsMenu' }>
          <Button size="mini" basic>
            <img src={ editIcon } alt="Edit Selection(s)" title="Edit Selection(s)" />
          </Button>

          <Button
            size="mini"
            basic
            onClick={ this.displayConfirmDelete }
          >
            <img src={ deleteIcon } alt="Delete Selection(s)" title="Delete Selection(s)" />
          </Button>

          <Mutation
            mutation={ DELETE_VIDEO_PROJECTS_MUTATION }
            onCompleted={ () => {
              this.props.handleResetSelections();
              this.handleDeleteCancel();
            } }
          >
            { ( deleteProjects, { error } ) => {
              if ( error ) return <ApolloError error={ error } />;

              return (
                <Confirm
                  className="delete"
                  open={ this.state.deleteConfirmOpen }
                  content={ (
                    <ConfirmModalContent
                      className="delete_confirm delete_confirm--video"
                      headline="Are you sure you want to deleteÂ the selected video project(s)?"
                    >
                      <p>The selected video project(s) will be permanently removed from the Content Cloud.</p>
                    </ConfirmModalContent>
                  ) }
                  onCancel={ this.handleDeleteCancel }
                  onConfirm={ () => this.handleDeleteConfirm( deleteProjects ) }
                  cancelButton="No, take me back"
                  confirmButton="Yes, delete forever"
                />
              );
            } }
          </Mutation>

          <Button size="mini" basic>
            <img src={ unpublishIcon } alt="Unpublish Selection(s)" title="Unpublish Selection(s)" />
          </Button>
          <Button size="mini" basic>
            <img src={ createIcon } alt="Create Selection(s)" title="Create Selection(s)" />
          </Button>
          <Button size="mini" basic>
            <img src={ archiveIcon } alt="Archive Selection(s)" title="Archive Selection(s)" />
          </Button>

          <span className="separator">|</span>

          <Mutation
            mutation={ UNPUBLISH_VIDEO_PROJECTS_MUTATION }
            onCompleted={ () => this.props.handleResetSelections() }
          >
            { ( unpublish, { error } ) => {
              if ( error ) return <ApolloError error={ error } />;
              return (
                <Button
                  className="unpublish"
                  size="mini"
                  basic
                  onClick={
                    () => this.handleUnpublish( unpublish )
                  }
                >
                  <span className="unpublish--text">Unpublish</span>
                </Button>
              );
            } }
          </Mutation>
        </div>
      </div>
    );
  }
}

TableActionsMenu.propTypes = {
  displayActionsMenu: PropTypes.bool,
  queryVariables: PropTypes.object,
  selectedItems: PropTypes.object,
  handleResetSelections: PropTypes.func,
  toggleAllItemsSelection: PropTypes.func
};

export default TableActionsMenu;
