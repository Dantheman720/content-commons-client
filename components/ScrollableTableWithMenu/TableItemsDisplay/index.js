/**
 *
 * TableItemsDisplay
 *
 */

import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import { Dropdown, Grid, Loader } from 'semantic-ui-react';
import ApolloError from 'components/errors/ApolloError';
import { TEAM_VIDEO_PROJECTS_COUNT_QUERY } from '../TablePagination';
import './TableItemsDisplay.scss';

const displaySizeOptions = [
  { key: 2, value: 2, text: '2' }, // temp value, for development
  { key: 4, value: 4, text: '4' }, // temp value, for development
  { key: 25, value: 25, text: '25' },
  { key: 50, value: 50, text: '50' },
  { key: 75, value: 75, text: '75' },
  { key: 100, value: 100, text: '100' },
];

const TableItemsDisplay = props => {
  const {
    handleChange,
    searchTerm,
    value: count,
    variables,
    variables: { skip }
  } = props;

  return (
    <Query
      query={ TEAM_VIDEO_PROJECTS_COUNT_QUERY }
      variables={ { ...variables } }
    >
      { ( { loading, error, data: { videoProjects } } ) => {
        if ( loading ) {
          return (
            <Grid.Column className="items_display">
              <Loader active inline size="mini" />
              <span>Loading...</span>
            </Grid.Column>
          );
        }
        if ( error ) {
          return (
            <Grid.Column className="items_display">
              <ApolloError error={ error } />
            </Grid.Column>
          );
        }
        if ( !videoProjects ) return null;

        const projectsCount = videoProjects.length;
        const firstPageItem = skip + 1;
        const range = skip + count;
        const lastPageItem = range < projectsCount ? range : projectsCount;

        return (
          <Grid.Column className="items_display">
            <span>
              Show:{ ' ' }
              <Dropdown
                inline
                options={ displaySizeOptions }
                value={ count }
                onChange={ ( e, { value } ) => handleChange( e, value ) }
              />
              { ' | ' }
              { projectsCount > 0
                ? <span>{ firstPageItem } - { lastPageItem } of { projectsCount }{ searchTerm && <Fragment> for &ldquo;{ searchTerm }&rdquo;</Fragment> }</span>
                : <span>No { searchTerm ? 'results' : 'projects' }{ searchTerm && <Fragment> for &ldquo;{ searchTerm }&rdquo;</Fragment> }</span> }
            </span>
          </Grid.Column>
        );
      } }
    </Query>
  );
};

TableItemsDisplay.propTypes = {
  handleChange: PropTypes.func,
  searchTerm: PropTypes.string,
  value: PropTypes.number,
  variables: PropTypes.object,
};

export default TableItemsDisplay;
