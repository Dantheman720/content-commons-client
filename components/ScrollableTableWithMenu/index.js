/**
 *
 * ScrollableTableWithMenu
 *
 */
import React from 'react';
import PropTypes from 'prop-types';
import sortBy from 'lodash/sortBy';
import debounce from 'lodash/debounce';
import { Table, Grid } from 'semantic-ui-react';
import { isMobile, isWindowWidthLessThanOrEqualTo } from 'lib/browser';
import TableHeader from './TableHeader';
import TableBody from './TableBody';
import TableItemsDisplay from './TableItemsDisplay';
import TableSearch from './TableSearch/TableSearch';
import TableMenu from './TableMenu';
import TableActionsMenu from './TableActionsMenu';
import TablePagination from './TablePagination';
import './ScrollableTableWithMenu.scss';

/* eslint-disable react/prefer-stateless-function */
class ScrollableTableWithMenu extends React.Component {
  state = {
    tableHeaders: this.props.persistentTableHeaders,
    selectedItems: new Map(),
    displayActionsMenu: false,
    column: null,
    direction: null,
    windowWidth: null,
    searchTerm: '',
    activePage: 1,
    itemsPerPage: 2, // set to low number for dev
    skip: 0
  };

  _selectAllItems = false;

  componentDidMount() {
    this.tableMenuSelectionsOnMobile();
    window.addEventListener( 'resize', this.tableMenuSelectionsOnResize );
  }

  componentWillUnmount() {
    window.removeEventListener( 'resize', this.tableMenuSelectionsOnResize );
  }

  handleItemsPerPageChange = ( e, value ) => {
    this.setState(
      { itemsPerPage: value, },
      this.handleResetActivePage
    );
  };

  handlePageChange = ( e, { activePage } ) => {
    this.setState( prevState => {
      const skip = activePage * prevState.itemsPerPage - prevState.itemsPerPage;
      return { activePage, skip };
    } );
  };

  handleResetActivePage = () => {
    this.setState( { activePage: 1, skip: 0 } );
  }

  handleResetSelections = () => {
    this.setState(
      {
        selectedItems: new Map(),
        displayActionsMenu: false
      },
      this.handleResetActivePage
    );
  }

  handleSearchSubmit = ( e, searchTerm ) => {
    e.preventDefault();
    this.setState(
      { searchTerm: searchTerm.trim() },
      this.handleResetActivePage
    );
  }

  handleSort = clickedColumn => () => {
    const {
      column,
      data,
      direction,
      displayActionsMenu
    } = this.state;

    if ( displayActionsMenu ) return;

    if ( column !== clickedColumn ) {
      return this.setState( {
        column: clickedColumn,
        data: sortBy( data, [clickedColumn] ),
        direction: 'ascending'
      } );
    }

    this.setState( {
      data: data.reverse(),
      direction: direction === 'ascending' ? 'descending' : 'ascending'
    } );
  };

  tableMenuOnChange = e => {
    e.persist();
    const menuItem = {
      name: e.target.parentNode.dataset.propname,
      label: e.target.parentNode.dataset.proplabel
    };
    this.setState( prevState => {
      if ( prevState.tableHeaders.map( h => h.name ).includes( menuItem.name ) ) {
        return {
          tableHeaders: prevState.tableHeaders.filter( h => h.name !== menuItem.name )
        };
      }
      return { tableHeaders: [...prevState.tableHeaders, menuItem] };
    } );
  };

  tableMenuSelectionsOnResize = () => {
    ( debounce( () => {
      const { persistentTableHeaders, columnMenu } = this.props;
      const windowWidth = window.innerWidth;
      const prevWindowWidth = this.state.windowWidth;
      if ( prevWindowWidth !== '' && prevWindowWidth <= 767 && !isWindowWidthLessThanOrEqualTo( 767 ) ) {
        return this.setState( { tableHeaders: persistentTableHeaders, windowWidth } );
      }
      if ( isWindowWidthLessThanOrEqualTo( 767 ) ) {
        return this.setState( { tableHeaders: [...persistentTableHeaders, ...columnMenu], windowWidth } );
      }
      return this.setState( { windowWidth } );
    }, 500, { leading: false, trailing: true } ) )();
  }

  tableMenuSelectionsOnMobile = () => {
    if ( isMobile() ) {
      const { columnMenu } = this.props;
      this.setState( prevState => ( {
        tableHeaders: [...prevState.tableHeaders, ...columnMenu]
      } ) );
    }
  }

  toggleAllItemsSelection = e => {
    e.stopPropagation();
    const allItems = Array
      .from( document.querySelectorAll( '[data-label]' ) )
      .map( item => item.dataset.label );

    this.setState( () => {
      const newSelectedItems = new Map();

      allItems.forEach( item => {
        if ( this._selectAllItems ) {
          newSelectedItems.set( item, false );
        } else {
          newSelectedItems.set( item, true );
        }
      } );

      this._selectAllItems = !this._selectAllItems;

      return ( {
        selectedItems: newSelectedItems,
        displayActionsMenu: this._selectAllItems
      } );
    } );
  }

  toggleItemSelection = ( e, data ) => {
    const isChecked = data.checked;
    this.setState( prevState => {
      const updatedSelectedItems = prevState.selectedItems.set( String( data['data-label'] ), isChecked );
      const areOtherItemsSelected = Array.from( updatedSelectedItems.values() ).includes( true );
      return {
        selectedItems: updatedSelectedItems,
        displayActionsMenu: areOtherItemsSelected
      };
    } );
  }

  render() {
    const {
      tableHeaders,
      displayActionsMenu,
      column,
      direction,
      selectedItems,
      searchTerm,
      activePage,
      itemsPerPage,
      skip,
      windowWidth
    } = this.state;

    const { columnMenu, team } = this.props;

    const variables = { team, searchTerm };

    const paginationVars = { first: itemsPerPage, skip };

    return (
      <Grid>
        <Grid.Row className="items_tableSearch">
          <TableSearch handleSearchSubmit={ this.handleSearchSubmit } />
        </Grid.Row>
        <Grid.Row className="items_tableMenus_wrapper">
          <Grid.Column mobile={ 16 } computer={ 3 }>
            <TableActionsMenu
              displayActionsMenu={ displayActionsMenu }
              variables={ { ...variables, ...paginationVars } }
              selectedItems={ selectedItems }
              handleResetSelections={ this.handleResetSelections }
              toggleAllItemsSelection={ this.toggleAllItemsSelection }
            />
          </Grid.Column>
          <Grid.Column mobile={ 16 } computer={ 13 } className="items_tableMenus">
            <TableItemsDisplay
              handleChange={ this.handleItemsPerPageChange }
              searchTerm={ searchTerm }
              value={ itemsPerPage }
              variables={ { ...variables, ...paginationVars } }
            />
            <TableMenu
              columnMenu={ columnMenu }
              tableMenuOnChange={ this.tableMenuOnChange }
            />
          </Grid.Column>
        </Grid.Row>
        <Grid.Row>
          <Grid.Column className="items_table_wrapper">
            <div className="items_table">
              <Table sortable celled>
                <TableHeader
                  tableHeaders={ tableHeaders }
                  column={ column }
                  direction={ direction }
                  handleSort={ this.handleSort }
                  toggleAllItemsSelection={ this.toggleAllItemsSelection }
                  displayActionsMenu={ displayActionsMenu }
                />
                <TableBody
                  searchTerm={ searchTerm }
                  selectedItems={ selectedItems }
                  tableHeaders={ tableHeaders }
                  toggleItemSelection={ this.toggleItemSelection }
                  variables={ { ...variables, ...paginationVars } }
                  windowWidth={ windowWidth }
                />
              </Table>
            </div>
          </Grid.Column>
        </Grid.Row>
        <Grid.Row>
          <Grid.Column className="items_tablePagination">
            <TablePagination
              activePage={ activePage }
              handlePageChange={ this.handlePageChange }
              itemsPerPage={ itemsPerPage }
              variables={ variables }
            />
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }
}

ScrollableTableWithMenu.propTypes = {
  persistentTableHeaders: PropTypes.array,
  columnMenu: PropTypes.array,
  team: PropTypes.string
};

export default ScrollableTableWithMenu;
