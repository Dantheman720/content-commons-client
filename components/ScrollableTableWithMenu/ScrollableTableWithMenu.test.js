import React from 'react';
import { shallow } from 'enzyme';

import ScrollableTableWithMenu from './ScrollableTableWithMenu';

describe( '<ScrollableTableWithMenu />', () => {
  it( 'renders without crashing', () => {
    shallow( <ScrollableTableWithMenu /> );
  } );
} );
