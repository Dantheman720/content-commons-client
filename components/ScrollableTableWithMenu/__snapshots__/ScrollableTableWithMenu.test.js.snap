// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScrollableTableWithMenu /> renders without crashing 1`] = `
<ScrollableTableWithMenu
  columnMenu={
    Array [
      Object {
        "label": "TEAM",
        "name": "team",
      },
      Object {
        "label": "CATEGORIES",
        "name": "categories",
      },
      Object {
        "label": "MODIFIED DATE",
        "name": "updatedAt",
      },
    ]
  }
  persistentTableHeaders={
    Array [
      Object {
        "label": "PROJECT TITLE",
        "name": "projectTitle",
      },
      Object {
        "label": "VISIBILITY",
        "name": "visibility",
      },
      Object {
        "label": "CREATED",
        "name": "createdAt",
      },
      Object {
        "label": "AUTHOR",
        "name": "author",
      },
    ]
  }
  projectTab="teamProjects"
  team="IIP Video Production"
>
  <Grid>
    <div
      className="ui grid"
    >
      <GridRow
        className="items_tableSearch"
      >
        <div
          className="row items_tableSearch"
        >
          <Component
            handleSearchSubmit={[Function]}
          >
            TableSearch
          </Component>
        </div>
      </GridRow>
      <GridRow
        className="items_tableMenus_wrapper"
      >
        <div
          className="row items_tableMenus_wrapper"
        >
          <GridColumn
            computer={3}
            mobile={16}
            tablet={3}
          >
            <div
              className="three wide computer sixteen wide mobile three wide tablet column"
            >
              <Component
                displayActionsMenu={false}
                handleResetSelections={[Function]}
                selectedItems={Array []}
                toggleAllItemsSelection={[Function]}
                variables={
                  Object {
                    "first": 15,
                    "searchTerm": "",
                    "skip": 0,
                    "team": "IIP Video Production",
                  }
                }
              >
                TableActionsMenu
              </Component>
            </div>
          </GridColumn>
          <GridColumn
            className="items_tableMenus"
            computer={13}
            mobile={16}
            tablet={13}
          >
            <div
              className="thirteen wide computer sixteen wide mobile thirteen wide tablet column items_tableMenus"
            >
              <Component
                handleChange={[Function]}
                searchTerm=""
                value={15}
                variables={
                  Object {
                    "first": 15,
                    "searchTerm": "",
                    "skip": 0,
                    "team": "IIP Video Production",
                  }
                }
              >
                TableItemsDisplay
              </Component>
              <Component
                columnMenu={
                  Array [
                    Object {
                      "label": "TEAM",
                      "name": "team",
                    },
                    Object {
                      "label": "CATEGORIES",
                      "name": "categories",
                    },
                    Object {
                      "label": "MODIFIED DATE",
                      "name": "updatedAt",
                    },
                  ]
                }
                tableMenuOnChange={[Function]}
              >
                TableMenu
              </Component>
            </div>
          </GridColumn>
        </div>
      </GridRow>
      <GridRow>
        <div
          className="row"
        >
          <GridColumn
            className="items_table_wrapper"
          >
            <div
              className="column items_table_wrapper"
            >
              <div
                className="items_table"
              >
                <Table
                  as="table"
                  celled={true}
                  sortable={true}
                >
                  <table
                    className="ui celled sortable table"
                  >
                    <TableHeader
                      column={null}
                      direction={null}
                      displayActionsMenu={false}
                      handleSort={[Function]}
                      tableHeaders={
                        Array [
                          Object {
                            "label": "PROJECT TITLE",
                            "name": "projectTitle",
                          },
                          Object {
                            "label": "VISIBILITY",
                            "name": "visibility",
                          },
                          Object {
                            "label": "CREATED",
                            "name": "createdAt",
                          },
                          Object {
                            "label": "AUTHOR",
                            "name": "author",
                          },
                        ]
                      }
                      toggleAllItemsSelection={[Function]}
                    >
                      <thead>
                        <tr>
                          <th>
                            TableHeader
                          </th>
                        </tr>
                      </thead>
                    </TableHeader>
                    <TableBody
                      column={null}
                      direction={null}
                      projectTab="teamProjects"
                      searchTerm=""
                      selectedItems={Array []}
                      tableHeaders={
                        Array [
                          Object {
                            "label": "PROJECT TITLE",
                            "name": "projectTitle",
                          },
                          Object {
                            "label": "VISIBILITY",
                            "name": "visibility",
                          },
                          Object {
                            "label": "CREATED",
                            "name": "createdAt",
                          },
                          Object {
                            "label": "AUTHOR",
                            "name": "author",
                          },
                        ]
                      }
                      toggleItemSelection={[Function]}
                      variables={
                        Object {
                          "first": 15,
                          "searchTerm": "",
                          "skip": 0,
                          "team": "IIP Video Production",
                        }
                      }
                      windowWidth={null}
                    >
                      <tbody>
                        <tr>
                          <td>
                            TableBody
                          </td>
                        </tr>
                      </tbody>
                    </TableBody>
                  </table>
                </Table>
              </div>
            </div>
          </GridColumn>
        </div>
      </GridRow>
      <GridRow>
        <div
          className="row"
        >
          <GridColumn
            className="items_tablePagination"
          >
            <div
              className="column items_tablePagination"
            >
              <Component
                activePage={1}
                handlePageChange={[Function]}
                itemsPerPage={15}
                variables={
                  Object {
                    "searchTerm": "",
                    "team": "IIP Video Production",
                  }
                }
              >
                TablePagination
              </Component>
            </div>
          </GridColumn>
        </div>
      </GridRow>
    </div>
  </Grid>
</ScrollableTableWithMenu>
`;
